subprojects/core/src/main/java/org/gradle/api/internal/tasks/properties/DefaultTypeMetadataStore.java

```diff
@@ -131,16 +141,23 @@ private static String calculateDisplayName(Iterable<? extends PropertyAnnotation
                 }
                 Class<? extends Annotation> annotationType = entry.getValue().annotationType();
                 if (!allowedModifiersForPropertyType.contains(annotationCategory)) {
-                    validationContext.visitPropertyProblem(WARNING,
-                        propertyAnnotationMetadata.getPropertyName(),
-                        String.format("is annotated with @%s that is not allowed for @%s properties",
-                            annotationType.getSimpleName(), propertyType.getSimpleName())
-                    );
+                    validationContext.visitPropertyProblem(problem ->
+                        problem.withId(ValidationProblemId.INCOMPATIBLE_ANNOTATIONS)
+                            .forProperty(propertyAnnotationMetadata.getPropertyName())
+                            .reportAs(ERROR)
+                            .withDescription(() -> "is annotated with @" + annotationType.getSimpleName() + " but that is not allowed for '" + propertyType.getSimpleName() + "' properties")
+                            .happensBecause(() -> "This modifier is used in conjunction with a property of type '" + propertyType.getSimpleName() + "' but this doesn't have semantics")
+                            .addPossibleSolution(() -> "Remove the '@" + annotationType.getSimpleName() + "' annotation")
+                            .documentedAt("validation_problems", "incompatible_annotations"));
                 } else if (!allowedPropertyModifiers.contains(annotationType)) {
-                    validationContext.visitPropertyProblem(ERROR,
-                        propertyAnnotationMetadata.getPropertyName(),
-                        String.format("has invalid annotation @%s",
-                            annotationType.getSimpleName())
+                    validationContext.visitPropertyProblem(problem ->
+                        problem.withId(ValidationProblemId.ANNOTATION_INVALID_IN_CONTEXT)
+                            .forProperty(propertyAnnotationMetadata.getPropertyName())
+                            .reportAs(ERROR)
+                            .withDescription(() -> String.format("has invalid annotation @%s", annotationType.getSimpleName()))
+                            .happensBecause(() -> "The '@" + propertyType.getSimpleName() + "' annotation cannot be used in this context")
+                            .addPossibleSolution("Remove the property")
```

Comment:
I think it would be great to list the allowed annotations for the context. Do you think that would make the message too long?

Reply:
Yes, it would. Actually I tried to so something simpler, which was replacing "in this context" with something more meaningful, like "in artifact transforms".
Unfortunately, the way the validation contexts are setup makes it difficult to pass this context. This includes the annotations which would make more sense.
I'd be keen on improving this if we have time.

Reply:
We can print `allowedPropertyModifiers`, right?